plugins/prometheus_plugin/prometheus_plugin.cpp

```diff
@@ -1,19 +1,168 @@
+         std::lock_guard<std::mutex> lock{collectables_mutex_};
+
+         auto locked = collectable.lock();
+         auto same_pointer = [&locked](const std::weak_ptr<Collectable>& candidate) {
+            return locked == candidate.lock();
+         };
+
+         collectables_.erase(std::remove_if(std::begin(collectables_),
+                                            std::end(collectables_), same_pointer),
+                             std::end(collectables_));
+      }
+
+      std::string scrape() {
+         auto start_time_of_request = std::chrono::steady_clock::now();
+
+         std::vector<MetricFamily> metrics;
+
+         {
+            std::lock_guard<std::mutex> lock{collectables_mutex_};
+            metrics = CollectMetrics(collectables_);
+         }
+
+         const TextSerializer serializer;
+
+         std::string body = serializer.Serialize(metrics);
+
+         auto stop_time_of_request = std::chrono::steady_clock::now();
+         auto duration = std::chrono::duration_cast<std::chrono::microseconds>(
+               stop_time_of_request - start_time_of_request);
+         request_latencies_.Observe(duration.count());
+
+         bytes_transferred_.Increment(body.length());
+         num_scrapes_.Increment();
+         return body;
+      }
+   };
+}
+
+#define CALL_WITH_400(api_name, api_handle, call_name, INVOKE, http_response_code) \
+{std::string("/v1/" #api_name "/" #call_name), \
+   [api_handle](string, string body, url_response_callback cb) mutable { \
+          try { \
+             body = parse_params<std::string, http_params_types::no_params>(body); \
+             INVOKE \
+             cb(http_response_code, fc::time_point::maximum(), fc::variant(result)); \
+          } catch (...) { \
+             http_plugin::handle_exception(#api_name, #call_name, body, cb); \
+          } \
+       }}
+
+#define INVOKE_R_V(api_handle, call_name) \
+     auto result = api_handle->call_name();
 
 namespace eosio {
 
    static appbase::abstract_plugin &_prometheus_plugin = app().register_plugin<prometheus_plugin>();
 
    struct prometheus_plugin_impl {
+      std::shared_ptr<prometheus::Registry> _prometheus_plugin_registry;
+      std::unique_ptr<prometheus::MetricsHandler> _metrics_handler;
+
+      prometheus_plugin_impl() :
+         _prometheus_plugin_registry(std::make_shared<prometheus::Registry>()),
+         _metrics_handler(std::make_unique<prometheus::MetricsHandler>(*_prometheus_plugin_registry)) {
+            register_collectable(_prometheus_plugin_registry);
+      }
+
+      void register_collectable(
+            const std::weak_ptr<prometheus::Collectable>& collectable) {
+         _metrics_handler->register_collectable(collectable);
+      }
+
+      std::string scrape() {return _metrics_handler->scrape();}
    };
 
    prometheus_plugin::prometheus_plugin() {
      my.reset(new prometheus_plugin_impl{});
+
+      app().get_plugin<http_plugin>().add_api({
```

Comment:
This can use `add_async_api`. `add_async_api` is a terrible name for that, but it registers it to run on the http_thread instead of the main thread, which is what we will want for this.